tfsdk/block.go

```diff
@@ -0,0 +1,412 @@
+func (b Block) terraformType(ctx context.Context) tftypes.Type {
+	return b.attributeType().TerraformType(ctx)
+}
+
+// tfprotov6 returns the *tfprotov6.SchemaNestedBlock equivalent of a Block.
+// Errors will be tftypes.AttributePathErrors based on `path`. `name` is the
+// name of the attribute.
+func (b Block) tfprotov6(ctx context.Context, name string, path *tftypes.AttributePath) (*tfprotov6.SchemaNestedBlock, error) {
+	schemaNestedBlock := &tfprotov6.SchemaNestedBlock{
+		Block: &tfprotov6.SchemaBlock{
+			Deprecated: b.DeprecationMessage != "",
+		},
+		MinItems: b.MinItems,
+		MaxItems: b.MaxItems,
+		TypeName: name,
+	}
+
+	if b.Description != "" {
+		schemaNestedBlock.Block.Description = b.Description
+		schemaNestedBlock.Block.DescriptionKind = tfprotov6.StringKindPlain
+	}
+
+	if b.MarkdownDescription != "" {
+		schemaNestedBlock.Block.Description = b.MarkdownDescription
+		schemaNestedBlock.Block.DescriptionKind = tfprotov6.StringKindMarkdown
+	}
+
+	nm := b.NestingMode
+	switch nm {
+	case NestingModeList:
+		schemaNestedBlock.Nesting = tfprotov6.SchemaNestedBlockNestingModeList
+	case NestingModeSet:
+		schemaNestedBlock.Nesting = tfprotov6.SchemaNestedBlockNestingModeSet
+	case NestingModeMap, NestingModeSingle:
+		// This is intentional to only maintain the previous Terraform Plugin SDK support.
+		return nil, path.NewErrorf("unsupported block nesting mode: %v", nm)
+	default:
+		return nil, path.NewErrorf("unrecognized nesting mode %v", nm)
+	}
+
+	for attrName, attr := range b.Attributes {
+		attrPath := path.WithAttributeName(attrName)
+		attrProto6, err := attr.tfprotov6SchemaAttribute(ctx, attrName, attrPath)
+
+		if err != nil {
+			return nil, err
+		}
+
+		schemaNestedBlock.Block.Attributes = append(schemaNestedBlock.Block.Attributes, attrProto6)
+	}
+
+	for blockName, block := range b.Blocks {
+		blockPath := path.WithAttributeName(blockName)
+		blockProto6, err := block.tfprotov6(ctx, blockName, blockPath)
+
+		if err != nil {
+			return nil, err
+		}
+
+		schemaNestedBlock.Block.BlockTypes = append(schemaNestedBlock.Block.BlockTypes, blockProto6)
+	}
+
+	sort.Slice(schemaNestedBlock.Block.Attributes, func(i, j int) bool {
+		if schemaNestedBlock.Block.Attributes[i] == nil {
+			return true
+		}
+
+		if schemaNestedBlock.Block.Attributes[j] == nil {
+			return false
+		}
+
+		return schemaNestedBlock.Block.Attributes[i].Name < schemaNestedBlock.Block.Attributes[j].Name
+	})
+
+	sort.Slice(schemaNestedBlock.Block.BlockTypes, func(i, j int) bool {
+		if schemaNestedBlock.Block.BlockTypes[i] == nil {
+			return true
+		}
+
+		if schemaNestedBlock.Block.BlockTypes[j] == nil {
+			return false
+		}
+
+		return schemaNestedBlock.Block.BlockTypes[i].TypeName < schemaNestedBlock.Block.BlockTypes[j].TypeName
+	})
+
+	return schemaNestedBlock, nil
+}
+
+// validate performs all Block validation.
+func (b Block) validate(ctx context.Context, req ValidateAttributeRequest, resp *ValidateAttributeResponse) {
```

Comment:
If I'm reading this right, we're opting _not_ to have block-level validation available. Is that the case? Is that by design?

Reply:
We can certainly add it, it's just not included yet. I don't think there's anything preventing its inclusion other than time/effort. If this is looking okay so far, I'll spend that time adding it.

(Aside: I also need to at least passthrough attribute plan modification through the block support still.)

Reply:
Addressed 👍